APIdefinition.md

```diff
@@ -0,0 +1,96 @@
+complete. The above philosophy should guide our thinking in what
+the public API is.
+
+
+## General Guidance
+
+The public API consists of the following.
+
+**Code-level structures**: Public function, member, and class names,
+argument signatures, and their return signatures.  Where "public"
+is defined as those elements that are typically provided for use external
+to a library or module, dependent on the language.
+
+Some examples are:
+
+- C++ public structures are those that can be access by including
+  the distributed header files for libraries.
+- Python public structures are those that do not begin with an underbar
+  ("_"), as described in PEP 8.
+
+
+**User-facing command-line and GUI programs**: program name, argument
+list, argument defaults, configuration or preference names and their
+defaults, and for those programs with structured output files, the
+structure of those output files are analogous to the return signatures
+of code-level structures.
+
+
+This guidance may be more straightforward to understand and apply
+for code-level structures which have more formalized argument and
+return signatures.  User-facing command-line and GUI programs should
+include verbose information in their documentation that describes what
+a user can expect as far as the structure of the output.
+
+
+## More Specific Examples
+
+Adding new arguments to code level structures does not violate the
+existing public API as long as they provide a default value. For
+example changing a function *doThing(int a, double b)* to *doThing(int
+a, double b, double c=1.0)* does not violate the existing API for
+the *doThing* function as code that calls the *doThing* function will
+still compile without modification.
+
+Adding new optional arguments or arguments with default values to
+command line programs does not violate the existing public API. For
+example, adding a new argument called "check" that has a default
+value of "false" to a command line program does not violate the
+program's public API as scripts and commands that were valid
+previously are still valid after the addition.
+
+
+### Unstructured text files
+
+For anything of an inherently unstructured file-type, like .txt;
+there is no guarnatee of its contents; just that it will exist.
+
+
+### CSV output files
+
+Column names and data types are part of the API.
+
+For example, a program that writes out a CSV should clearly describe
+the output columns which include the fieldname, data type, and
+whether non-data-type values can be present (e.g. a column of
+floating point numbers named "minimum" that could contain the string
+"NULL" instead of a floating point value).
+
+New columns can be added (feature), but if a column's characteristics are 
+changed or removed, that's a breaking change.
+
+
+### PVL files
+
+Object, group, and keyword names are a part of the API.
+
+For example, the "minimum" keyword will exist in the "statistics"
+group. This also applies to the PVL labels in Cube files.
+
```

Comment:
I think that's still breaking. From a development perspective it is definitely a bug; but from a user perspective, any programmatic tools that interface with the file will need to be coded against that bug. So, fixing the issue will break any tool that references that column by name. So, I think from the perspective of breaking changes requiring modifications to existing consumers, this would be breaking.

Reply:
Based on [this discussion](https://github.com/planetarypy/pvl/issues/99), it seems that the perspective of the library can and should supersede the perspective of the user in this case.

Reply:
I think that's a philosophical decision we need to make. If we do take the stand this such a change is a bug fix we need to be very careful about how we communicate it to users on release. Basically, say "this change is a bug fix but it has a chance to break existing tools if you coded against the previously faulty behavior." We likely should also call it out in a specific section of the release notes/changelog.

Reply:
This seems like a discussion to have on a particular PR that might then influence an example being added to this document.  Are there specific changes that need to be made to *this* document, or can this be resolved?